### PR TITLE
[USMON-782] usm: http2: Dynamic table cleaning mechanism

### PR DESCRIPTION
**Pull request by guyarb**
### What does this PR do?

Introduces a new probe to cleanup the dynamic table and ensure no excess entries remains in the map.

### Motivation

During dogfooding we found out the dynamic table is being filled, and the cleanup happens only during
closing of the connection, so long living connection with many requests, can cause the map to 
be filled up to the limit, which results in a drop in the accuracy.

### Additional Notes

We introduce a new pair of probes per connenction (one for plaintext and one for TLS) to handle the cleanup
of the dynamic table.

We maintain 2 pointers per connection:
  1. "current" - dynamic table counter (the existing value we have).
  2. "previous" - used to store the latest entry we've clean.

After processing the headers, the cleanup phase kicks off. If "current" - "previous" reaches the threshold,
 we're cleaning all 'evicted' entries.

HTTP2 dynamic table is 4KB by default, but subject to changes per connection. Each entry consists of
  - the length of the key
  - the length of the value
  - and a constant of 32B.
(source - https://datatracker.ietf.org/doc/html/rfc7541#section-4.1)
If we assume all keys and values have a length of 1B, then we get a lower bound of 34B per
entry, which means we can store up to 120 entries (4KB / 32B) in the dynamic table before the eviction
happens.

Caveats:
  1. Since we don't really know what entries have been inserted to the dynamic table, we assume all
     of them were inserted, and we try to delete every entry in a specific range, hence there is at least 1 map
     operation overhead.
  2. We set a threshold of 200 entries, to support cases where the dynamic table size is bigger
     than 4KB. That means if we have more than 200 entries, any subsequent entry after 200th will be cleaned.
     For example, having the dynamic global counter pointing to 205, means we'll clean the entries 1 to 5.
  3. While we set the threshold for triggering the cleanup to 200, we actually can delete up to
     300 entries, to cover edge cases and gaps. For example in case a new frame introduces multiple new literal
     headers, and we cross the 200 threshold. For example, the difference is 198 entries, and a new frame
     introduces 5 new entries, thus the diff now is 203, which triggers the cleanup but misses 3 evicted entries.

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 1]
```diff
@@ -860,6 +860,43 @@ int socket__http2_headers_parser(struct __sk_buff *skb) {
     return 0;
 }
 
+SEC("socket/http2_dynamic_table_cleaner")
```
Comment (amitslavin) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:863]: After getting the dispatcher_args_copy the all logic is the same, (except of the next call). Should we move it the common to reduce code duplication?

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 2]
```diff
@@ -860,6 +860,43 @@ int socket__http2_headers_parser(struct __sk_buff *skb) {
...
+    if (!fetch_dispatching_arguments(&dispatcher_args_copy.tup, &dispatcher_args_copy.skb_info)) {
+        return 0;
+    }
+
+    dynamic_counter_t *dynamic_counter = bpf_map_lookup_elem(&http2_dynamic_counter_table, &dispatcher_args_copy.tup);
+    if (dynamic_counter == NULL) {
+        goto next;
+    }
+
+    if (dynamic_counter->value - dynamic_counter->previous <= HTTP2_DYNAMIC_TABLE_CLEANUP_THRESHOLD) {
```
Comment (amitslavin) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:876]: add documentation

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 3]
```diff
@@ -860,6 +860,43 @@ int socket__http2_headers_parser(struct __sk_buff *skb) {
...
+    };
+
+    #pragma unroll(HTTP2_DYNAMIC_TABLE_CLEANUP_ITERATIONS)
+    for (__u16 index = 0; index < HTTP2_DYNAMIC_TABLE_CLEANUP_ITERATIONS; index++) {
+        if (dynamic_counter->previous + HTTP2_DYNAMIC_TABLE_CLEANUP_THRESHOLD >= dynamic_counter->value) {
+            break;
+        }
+        dynamic_index.index = dynamic_counter->previous;
+        bpf_map_delete_elem(&http2_dynamic_table, &dynamic_index);
+        dynamic_counter->previous++;
```
Comment (amitslavin) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:891]: add documentation over here as well to make it more clear by reading the code

**Review by amitslavin — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/protocols/http2/decoding.h (hunk 1 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/http2/decoding.h:863]: I think that's something we should do, but no in the scope of this PR

**Review by amitslavin — APPROVED** [review]

**Pull request merged**